src/volatile_memory.rs

```diff
@@ -235,6 +237,55 @@ impl<'a> From<&'a mut [u8]> for VolatileSlice<'a, ()> {
 #[repr(C, packed)]
 struct Packed<T>(T);
 
+/// A slice of mapped memory that supports volatile access.
+pub struct VolatileSliceGuard {
+    slice: &'static mut [u8],
+
+    // This isn't used directly by any helper routine, but this protects the slice from getting
+    // unmapped while in use. Once this goes out of scope, the memory is automatically unmapped.
+    #[allow(dead_code)]
+    #[cfg(all(feature = "xen", unix))]
+    mmap_slice: crate::MmapSlice,
+}
+
+impl VolatileSliceGuard {
+    #[cfg(not(all(feature = "xen", unix)))]
+    fn new(_mmap: Option<&MmapInfo>, addr: *mut u8, len: usize, _is_read: bool) -> Self {
+        Self {
+            // SAFETY: safe because this is built from an already available slice.
+            slice: unsafe { slice::from_raw_parts_mut(addr, len) },
+        }
+    }
+
+    #[cfg(all(feature = "xen", unix))]
+    fn new(mmap: Option<&MmapInfo>, addr: *mut u8, len: usize, is_read: bool) -> Self {
+        let mmap_slice = mmap_slice(
+            mmap,
+            addr,
+            len,
+            if is_read {
+                libc::PROT_READ
+            } else {
+                libc::PROT_WRITE
+            },
+        );
+
+        Self {
+            // SAFETY: safe because this is built from an already available slice.
+            slice: unsafe { slice::from_raw_parts_mut(mmap_slice.addr(), len) },
```

Comment:
But who guards us from mapping this region twice? That would cause unsafety, right?

Also, I think this is unsafe because we are still dealing with volatile memory that a guest may write to (even if it is forbidden to do) which would probably be undefined behaviour on the Rust side. VolatileSlice seems to dance around this carefully by not treating the memory as slice but only offering to copy it to and from slices (at leasts that was my understanding when I last reviewed the code). So probably something similar would need to be done here (while also giving access to the raw pointer for zero-copy mechanisms). In fact, I would argue that if you call this API, the convenience functions of VolatileSlice are not enough for you (since you are using Xen and want access to the pointer for zero-copy read/writes). So maybe we can just get rid of the slice completely and only offer a pointer?

Reply:
Discussed with @Ablu offline and decided to remove the slice support and rather expose the pointer directly in the new implementation.

@Ablu please resolve this comment if you are satisfied with the new patches.

Reply:
LGTM (cannot mark as resolved since I commented on a commit that no longer exists)